src/main/java/uk/gov/nationalarchives/pdi/step/xml/CanonicalStep.java

```diff
@@ -0,0 +1,121 @@
+package uk.gov.nationalarchives.pdi.step.xml;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.xml.security.c14n.Canonicalizer;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.row.RowDataUtil;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.*;
+import org.w3c.dom.Document;
+import org.xml.sax.InputSource;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import java.io.ByteArrayOutputStream;
+import java.io.StringReader;
+
+public class CanonicalStep extends BaseStep implements StepInterface {
+    /**
+     * This is the base step that forms that basis for all steps. You can derive from this class to implement your own
+     * steps.
+     *
+     * @param stepMeta          The StepMeta object to run.
+     * @param stepDataInterface the data object to store temporary data, database connections, caches, result sets,
+     *                          hashtables etc.
+     * @param copyNr            The copynumber for this step.
+     * @param transMeta         The TransInfo of which the step stepMeta is part of.
+     * @param trans             The (running) transformation to obtain information shared among the steps.
+     */
+    public CanonicalStep(StepMeta stepMeta, StepDataInterface stepDataInterface, int copyNr, TransMeta transMeta, Trans trans) {
+        super(stepMeta, stepDataInterface, copyNr, transMeta, trans);
+    }
+
+    @Override
+    public boolean processRow(final StepMetaInterface smi, final StepDataInterface sdi) throws KettleException {
+        final CanonicalStepMeta meta = (CanonicalStepMeta) smi;
+        final CanonicalStepData data = (CanonicalStepData) sdi;
+
+        final Object[] row = getRow();
+        if (row == null) {
+            setOutputDone();
+            return false;
+        }
+
+        if(first) {
+            first = false;
+            data.setOutputRowMeta(getInputRowMeta().clone());
+            meta.getFields( data.getOutputRowMeta(), getStepname(), null, null, this, null, null );
```

Comment:
Minor thing - Can you run IntelliJ code format over this please. I see some spaces after braces in some places but not others.